File: flutter_app/lib/models/stock_history.dart
import 'package:json_annotation/json_annotation.dart';

part 'stock_history.g.dart';

enum StockMovementType {
  @JsonValue('INITIAL')
  initial,
  @JsonValue('RESTOCK')
  restock,
  @JsonValue('SALE')
  sale,
  @JsonValue('ADJUSTMENT')
  adjustment,
  @JsonValue('RETURN')
  returnItem,
  @JsonValue('DAMAGE')
  damage,
  @JsonValue('TRANSFER')
  transfer,
}

@JsonSerializable()
class StockHistory {
  @JsonKey(name: '_id')
  final String id;
  final String menuItemId;
  final String userId;
  final StockMovementType movementType;
  final int quantityChange;
  final int previousQuantity;
  final int newQuantity;
  final String? reason;
  final String? saleId;
  final String? performedBy;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const StockHistory({
    required this.id,
    required this.menuItemId,
    required this.userId,
    required this.movementType,
    required this.quantityChange,
    required this.previousQuantity,
    required this.newQuantity,
    this.reason,
    this.saleId,
    this.performedBy,
    this.createdAt,
    this.updatedAt,
  });

  factory StockHistory.fromJson(Map<String, dynamic> json) {
    // Handle MongoDB ObjectId conversions
    Map<String, dynamic> processedJson = Map<String, dynamic>.from(json);

    // Convert _id
    if (processedJson.containsKey('_id') && processedJson['_id'] is Map) {
      processedJson['_id'] = processedJson['_id']['\$oid'];
    }

    // Convert menuItemId
    if (processedJson.containsKey('menuItemId') &&
        processedJson['menuItemId'] is Map) {
      processedJson['menuItemId'] = processedJson['menuItemId']['\$oid'];
    }

    // Convert userId
    if (processedJson.containsKey('userId') && processedJson['userId'] is Map) {
      processedJson['userId'] = processedJson['userId']['\$oid'];
    }

    // Convert saleId
    if (processedJson.containsKey('saleId') && processedJson['saleId'] is Map) {
      processedJson['saleId'] = processedJson['saleId']['\$oid'];
    }

    // Convert performedBy
    if (processedJson.containsKey('performedBy') &&
        processedJson['performedBy'] is Map) {
      processedJson['performedBy'] = processedJson['performedBy']['\$oid'];
    }

    // Convert dates
    if (processedJson.containsKey('createdAt')) {
      if (processedJson['createdAt'] is Map &&
          processedJson['createdAt'].containsKey('\$date')) {
        processedJson['createdAt'] = processedJson['createdAt']['\$date'];
      }
    }

    if (processedJson.containsKey('updatedAt')) {
      if (processedJson['updatedAt'] is Map &&
          processedJson['updatedAt'].containsKey('\$date')) {
        processedJson['updatedAt'] = processedJson['updatedAt']['\$date'];
      }
    }

    return _$StockHistoryFromJson(processedJson);
  }

  Map<String, dynamic> toJson() => _$StockHistoryToJson(this);
}

// Extension for display strings
extension StockMovementTypeExtension on StockMovementType {
  String get displayName {
    switch (this) {
      case StockMovementType.initial:
        return 'Initial Stock';
      case StockMovementType.restock:
        return 'Restocked';
      case StockMovementType.sale:
        return 'Sale';
      case StockMovementType.adjustment:
        return 'Adjustment';
      case StockMovementType.returnItem:
        return 'Return';
      case StockMovementType.damage:
        return 'Damage/Expired';
      case StockMovementType.transfer:
        return 'Transfer';
    }
  }

  String get icon {
    switch (this) {
      case StockMovementType.initial:
        return '📦';
      case StockMovementType.restock:
        return '📥';
      case StockMovementType.sale:
        return '💰';
      case StockMovementType.adjustment:
        return '⚙️';
      case StockMovementType.returnItem:
        return '↩️';
      case StockMovementType.damage:
        return '❌';
      case StockMovementType.transfer:
        return '🔄';
    }
  }
}

File: business-sales-backend/src/models/StockHistory.ts
import mongoose, { Schema, Document, type InferSchemaType } from 'mongoose';

// Stock movement types
export enum StockMovementType {
  INITIAL = 'INITIAL',           // Initial stock entry
  RESTOCK = 'RESTOCK',           // Adding new stock
  SALE = 'SALE',                 // Stock sold
  ADJUSTMENT = 'ADJUSTMENT',     // Manual adjustment
  RETURN = 'RETURN',             // Item returned
  DAMAGE = 'DAMAGE',             // Damaged/expired items
  TRANSFER = 'TRANSFER'          // Stock transfer
}

const StockHistorySchema = new Schema({
  menuItemId: {
    type: Schema.Types.ObjectId,
    required: true,
    ref: 'MenuItems',
    index: true
  },
  userId: {
    type: Schema.Types.ObjectId,
    required: true,
    ref: 'Users',
    index: true
  },
  movementType: {
    type: String,
    enum: Object.values(StockMovementType),
    required: true
  },
  quantityChange: {
    type: Number,
    required: true,
    validate: {
      validator: function(value: number) {
        return Number.isInteger(value) && value !== 0;
      },
      message: 'Quantity change must be a non-zero integer'
    }
  },
  previousQuantity: {
    type: Number,
    required: true,
    min: 0
  },
  newQuantity: {
    type: Number,
    required: true,
    min: 0
  },
  reason: {
    type: String,
    trim: true,
    maxlength: 200
  },
  saleId: {
    type: Schema.Types.ObjectId,
    ref: 'SaleRecords',
    required: false // Only for SALE movements
  },
  performedBy: {
    type: Schema.Types.ObjectId,
    ref: 'Users',
    required: false // Can be null for automated movements
  }
}, {
  timestamps: true
});

// Generate the type from the schema
type StockHistorySchemaType = InferSchemaType<typeof StockHistorySchema>;

// Create the interface extending Document and the schema type
interface IStockHistory extends StockHistorySchemaType, Document {}

// Indexes for better query performance
StockHistorySchema.index({ menuItemId: 1, createdAt: -1 }); // Get history for an item
StockHistorySchema.index({ userId: 1, createdAt: -1 }); // Get user's stock history
StockHistorySchema.index({ movementType: 1 }); // Filter by movement type
StockHistorySchema.index({ createdAt: -1 }); // Sort by date

export default mongoose.model<IStockHistory>('StockHistory', StockHistorySchema, 'StockHistory');

